util/git/client.go

```diff
@@ -385,20 +396,87 @@ func (m *nativeGitClient) LsRefs() (*Refs, error) {
 // runs with in-memory storage and is safe to run concurrently, or to be run without a git
 // repository locally cloned.
 func (m *nativeGitClient) LsRemote(revision string) (res string, err error) {
+	set := make(map[subModule]bool)
+	var submoduleRefs []string
+	if os.Getenv(common.EnvGitSubmoduleUpdateChanges) == "true" {
+		gitmodules, err := m.loadGitmodulesFile()
+		if err == nil {
+			scanner := bufio.NewScanner(gitmodules)
+			var subModules []subModule
+			for scanner.Scan() {
+				line := strings.TrimSpace(scanner.Text())
+				re := regexp.MustCompile(`\[\s*submodule.*\]`)
+				if re.MatchString(line) {
+					log.Info("Get Infos for submodule: " + line)
+					// new submodule starts
+					if len(subModules) > 0 {
+						set[subModules[0]] = true
+						subModules = nil
+					}
+					module := subModule{
+						URL:    "",
+						Branch: "master",
+					}
+					subModules = append(subModules, module)
+				}
+
+				if strings.HasPrefix(line, "url") {
+					subModules[0].URL = m.extractSubmoduleURL(line)
+				}
+				if strings.HasPrefix(line, "branch") {
+					subModules[0].Branch = m.extractSubmoduleBranch(line)
+				}
+			}
+
+			submoduleRefs = m.getSubmoduleSHAs(set, submoduleRefs)
+		}
+		defer gitmodules.Close()
+	}
 	for attempt := 0; attempt < maxAttemptsCount; attempt++ {
-		if res, err = m.lsRemote(revision); err == nil {
+		if res, err = m.lsRemote(revision, ""); err == nil && len(submoduleRefs) > 0 {
+			if len(submoduleRefs) > 0 {
+				res = res + "-"
+			}
+			for _, submoduleRef := range submoduleRefs {
+				res = res + submoduleRef
+			}
 			return
 		}
 	}
 	return
 }
 
-func (m *nativeGitClient) lsRemote(revision string) (string, error) {
-	if IsCommitSHA(revision) {
-		return revision, nil
+func (m *nativeGitClient) loadGitmodulesFile() (*os.File, error) {
+	return os.Open(m.root + "/.gitmodules")
+}
+
+func (m *nativeGitClient) extractSubmoduleURL(line string) string {
+	re := regexp.MustCompile(`url\s*=\s*`)
+	urlPrefix := re.FindString(line)
+	return strings.TrimPrefix(line, urlPrefix)
+}
+
+func (m *nativeGitClient) extractSubmoduleBranch(line string) string {
+	re := regexp.MustCompile(`branch\s*=\s*`)
+	branchPrefix := re.FindString(line)
+	return strings.TrimPrefix(line, branchPrefix)
+}
+
+func (m *nativeGitClient) getSubmoduleSHAs(set map[subModule]bool, submoduleRefs []string) []string {
+	for key := range set {
+		ref, _ := m.lsRemote(key.Branch, key.URL)
+		submoduleRefs = append(submoduleRefs, ref)
+	}
+	return submoduleRefs
+}
+
+func (m *nativeGitClient) lsRemote(revision string, moduleURL string) (string, error) {
+	commitSHACorrected := strings.Split(revision, "-")
+	if len(commitSHACorrected) > 0 && IsCommitSHA(commitSHACorrected[0]) && moduleURL == "" {
```

Comment:
`len(commitSHACorrected)` is always > 0 after using `strings.Split()` with `sep` being set.